src/services/relay_renewal_job/refresh_topic_subscriptions.rs

```diff
@@ -0,0 +1,85 @@
+use {
+    crate::{
+        error::NotifyServerError,
+        metrics::Metrics,
+        model::helpers::{get_project_topics, get_subscriber_topics},
+    },
+    futures_util::{StreamExt, TryFutureExt, TryStreamExt},
+    relay_client::http::Client,
+    relay_rpc::{domain::Topic, rpc::MAX_SUBSCRIPTION_BATCH_SIZE},
+    sqlx::PgPool,
+    std::time::Instant,
+    tracing::{info, instrument},
+    wc::metrics::otel::Context,
+};
+
+// TODO change error type
+#[instrument(skip_all)]
+pub async fn run(
+    key_agreement_topic: Topic,
+    client: &Client,
+    postgres: &PgPool,
+    metrics: Option<&Metrics>,
+) -> Result<(), NotifyServerError> {
+    // TODO only renew when the subscription needs it
+    info!("Resubscribing to all topics");
+    let start = Instant::now();
+
+    let subscriber_topics = get_subscriber_topics(postgres, metrics).await?;
+    let subscriber_topics_count = subscriber_topics.len();
+    info!("subscriber_topics_count: {subscriber_topics_count}");
+
+    let project_topics = get_project_topics(postgres, metrics).await?;
+    let project_topics_count = project_topics.len();
+    info!("project_topics_count: {project_topics_count}");
+
+    // TODO: These need to be paginated and streamed from the database directly
+    // instead of collecting them to a single giant vec.
+    let topics = [key_agreement_topic]
+        .into_iter()
+        .chain(subscriber_topics.into_iter())
+        .chain(project_topics.into_iter())
+        .collect::<Vec<_>>();
+    let topics_count = topics.len();
+    info!("topics_count: {topics_count}");
+
+    // Collect each batch into its own vec, since `batch_subscribe` would convert
+    // them anyway.
+    let topics = topics
+        .chunks(MAX_SUBSCRIPTION_BATCH_SIZE)
+        .map(|chunk| chunk.to_vec())
+        .collect::<Vec<_>>();
+
+    // Limit concurrency to avoid overwhelming the relay with requests.
+    const REQUEST_CONCURRENCY: usize = 200;
+
+    futures_util::stream::iter(topics)
+        .map(|topics| {
+            // Map result to an unsized type to avoid allocation when collecting,
+            // as we don't care about subscription IDs.
+            client.batch_subscribe_blocking(topics).map_ok(|_| ())
+        })
+        .buffer_unordered(REQUEST_CONCURRENCY)
```

Comment:
Can you please make sure that concurrency of all relay calls are limited to some sane number? The reason is because notify is not subject to regular rate limits, and we need to be careful with the load. Although HTTP _should in theory_ distribute the load between multiple relay nodes, my testing showed that that's not always the case.

Reply:
The concurrency is constrained by two things:
- The rate limits from clients of Notify Server
- The concurrency limits of background jobs, such as this one here

I haven't done the math on what this results to, but endpoints allow no more than 1 req/s per client on average and currently we have about 1 req/s average from usage so we should be in the green.

Having a more global rate limiting mechanism such that we limit the absolute rate of requests across the entire service will be a complex refactor, which I would have to think more about to know how to approach. But in general this is still flawed as we being to implement horizontal scaling where we can't have such a global limit. If relay is having problems processing the request load, then relay becomes a bottleneck to the user-facing performance of Notify Server so performance must be improved in relay to support.